Documentation/contributing/development/statedb.rst

```diff
@@ -0,0 +1,379 @@
+
+* By default publicly provide `Table[Obj]` so new features can build on it and it can be used in tests. Also export the table's indexes or the query functions (`var ByName = nameIndex.Query`).
+
+* Do not export `RWTable[Obj]` if outside modules do not need to directly write into the table. If other modules do write into the table, consider defining "writer functions" that validate that the writes are well-formed.
+
+* If the table is closely associated with a specific feature, define it alongside the
+implementation of the feature. If the table is shared by many modules, consider defining it in `daemon/k8s` or `pkg/datapath/tables` so it is easy to discover.
+
+* Make sure the object is JSON marshallable so it can be inspected. If you need to store non-marshallable data (e.g. functions), make them private or mark them with `json:"-"`.
+
+* If the object contains a map or set and it is often mutated, consider using the immutable `part.Map` or `part.Set` from `cilium/statedb`. Since these are immutable they don't need to be deep-copied when modifying the object and there's no risk of accidentally mutating them in-place.
+
+* When designing a table consider how it can be used in tests outside your module. It's a good idea to export your table constructor (New*Table) so it can be used by itself in an integration test of a module that depends on it.
+
+* Take into account the fact that objects be immutable by designing them to be cheap to shallow-clone. For example this could mean splitting off fields that are constant from creation into their own struct that's referenced from the object.
+
+* Write benchmarks for your table to understand the cost of the indexing and storage use. See `benchmarks_test.go` in `cilium/statedb` for examples.
+
+* If the object is small (<100 bytes) prefer storing it by value instead of by reference, e.g. `Table[MyObject]` instead of `Table[*MyObject]`. This reduces memory fragmentation and makes it safer to use since the fields can't be accidentally mutated (anything inside that's by reference of course can be mutated accidentally). Note though that each index will store a separate copy of the object. Measure if needed.
+
+With that out of the way, let's get concrete with a code example of a simple
+table and a controller that populates it:
+
+FIXME: put the code into files that we test and include it here.
+
+.. code-block:: go
+
+	// Example is our object that we want to index and store in a table.
+	type Example struct {
+		ID uint64
+		Label string
+	}
+
+	// TableHeader defines how cilium-dbg displays the header
+	func (e *Example) TableHeader() []string {
+		return []string{
+			"ID",
+			"Label",
+		}
+	}
+
+	// TableRow defines how cilium-dbg displays a row
+	func (e *Example) TableRow() []string {
+		return []string{
+			strconv.FormatUint(e.ID, 10, 64),
+			e.Label,
+		}
+	}
+
+	// TableName is a constant for the table name. This is used in cilium-dbg
+	// to refer to this table.
```

Comment:
```suggestion
	// to refer to this table. The TableName must be composed of lowercase letters, separated by hyphens, and its total length cannot exceed X characters.
```
TODO, we should define a maximum limit of characters we can set on a table

Reply:
please review: https://github.com/cilium/statedb/pull/52